### PR TITLE
fsmonitor: implement filesystem change listener for Linux

### PR DESCRIPTION
**Pull request by ptarjan**
This series implements the built-in fsmonitor daemon for Linux using the inotify API, bringing it to feature parity with the existing Windows and macOS implementations. It also fixes two memory leaks in the platform-independent daemon code and deduplicates the IPC and settings logic that is now shared between macOS and Linux.

The implementation uses inotify rather than fanotify because fanotify requires either CAP_SYS_ADMIN or CAP_PERFMON capabilities, making it unsuitable for an unprivileged user-space daemon. While inotify has the limitation of requiring a separate watch on every directory (unlike macOS FSEvents, which can monitor an entire directory tree with a single watch), it operates without elevated privileges and provides the per-file event granularity needed for fsmonitor.

The listener uses inotify_init1(O_NONBLOCK) with a poll loop that checks for events with a 50-millisecond timeout, keeping the inotify queue well-drained to minimize the risk of overflows. Bidirectional hashmaps map between watch descriptors and directory paths for efficient event resolution. Directory renames are tracked using inotify cookie mechanism to correlate IN_MOVED_FROM and IN_MOVED_TO event pairs; a periodic check detects stale renames where the matching IN_MOVED_TO never arrived, forcing a resync.

New directory creation triggers recursive watch registration to ensure all subdirectories are monitored. The IN_MASK_CREATE flag is used where available to prevent modifying existing watches, with a fallback for older kernels. When IN_MASK_CREATE is available and inotify_add_watch returns EEXIST, it means another thread or recursive scan has already registered the watch, so it is safe to ignore.

Remote filesystem detection uses statfs() to identify network-mounted filesystems (NFS, CIFS, SMB, FUSE, etc.) via their magic numbers. Mount point information is read from /proc/mounts and matched against the statfs f_fsid to get accurate, human-readable filesystem type names for logging. When the .git directory is on a remote filesystem, the IPC socket falls back to $HOME or a user-configured directory via the fsmonitor.socketDir setting.

This series builds on work from https://github.com/git/git/pull/1352 by Eric DeCosta and https://github.com/git/git/pull/1667 by Marziyeh Esipreh, updated to work with the current codebase and address all review feedback.

Changes since v14:
- Fix stop_daemon_delete_repo: test_might_fail doesn't allow maybe_timeout, replaced with `{ ... || :; }` per SZEDER/Peff review, restoring broken &&-chain

Changes since v13:
- Also disable GIT_TEST_SPLIT_INDEX in t9211 per Junio's feedback (every test does scalar clone)

Changes since v12:
- Dropped both the fsmonitor.c workaround and the read-cache.c skipHash fix per Dscho's review: split-index and index.skipHash are fundamentally incompatible
- Added sane_unset GIT_TEST_SPLIT_INDEX to scalar clone tests in t9210 (patch 1/13)

Changes since v11:
- Fix t9210 BUG assertion with GIT_TEST_SPLIT_INDEX=yes: guard fsmonitor_dirty bitmap access against split-index having fewer entries than the bitmap expects

Changes since v10:
- Reverted pre_exec_cb callback back to simple close_fd_above_stderr flag per Junio's clarification (same as v8)

Changes since v9:
- Fixed Windows build: close_fd_above_stderr() compiles as a no-op on Windows since there is no fork/exec

Changes since v8:
- Replaced close_fd_above_stderr flag with generic pre_exec_cb callback in struct child_process per Junio's review, with close_fd_above_stderr() as a ready-made callback

Changes since v7:
- Added patch 12: convert khash to strset in do_handle_client (Patrick's #leftoverbit suggestion)
- Fixed "Forcing shutdown" trace message to start with lowercase
- Fixed redundant statfs() call in find_mount() (caller already had the result)
- Fixed CMakeLists.txt GIT-BUILD-OPTIONS: was hardcoded to "win32" for FSMONITOR_DAEMON_BACKEND and FSMONITOR_OS_SETTINGS, now uses the CMake variables
- Fixed uninitialized strset on trivial response path (STRSET_INIT)
- Removed V9FS_MAGIC from get_fs_typename() to match is_remote_fs() (9p is local VM mounts)
- Split 30-second stop timeout into its own commit per review request
- Fixed misleading indentation on shutdown assignment in handle_events()
- Updated commit messages to describe all changes (test hardening, fsmonitor-ipc.c spawn changes)
- Updated Makefile comment for FSMONITOR_OS_SETTINGS to mention fsm-ipc

Changes since v6:
- Introduced FSMONITOR_OS_SETTINGS build variable (set to "unix" for macOS and Linux, "win32" for Windows) to eliminate if/else conditionals in Makefile, meson.build, and CMakeLists.txt per Junio's review
- Moved fsm-path-utils from FSMONITOR_OS_SETTINGS to FSMONITOR_DAEMON_BACKEND since path-utils files are platform-specific
- Removed V9FS_MAGIC from remote filesystem detection (9p is used for local VM/container host mounts where fsmonitor works fine)
- Removed redundant #include <libgen.h> (already provided by compat/posix.h)
- Fixed cookie wait comment wording ("see" to "observe")
- Rewrote commit messages for IPC and settings dedup patches

Changes since v5:
- Split monolithic commit into 10-patch series per Patrick's review
- Deduplicated fsm-ipc and fsm-settings into shared Unix implementations
- Rewrote commit message with prose paragraphs, explain inotify vs fanotify, removed "Issues addressed" sections, added Based-on-patch-by trailers
- Removed redundant includes already provided by compat/posix.h
- Fixed error/trace message capitalization per coding guidelines
- Fixed stale rename check interval from 1000 seconds to 1 second
- Changed poll timeout from 1ms to 50ms to reduce idle CPU wake-ups
- Replaced infinite pthread_cond_wait cookie loop with one-second pthread_cond_timedwait (prevents daemon hangs on overlay filesystems where events are never delivered)
- Added pthread_cond_timedwait to Windows pthread compatibility layer
- Separated test into its own commit with smoke test that skips when inotify events are not delivered (e.g., overlayfs with older kernels)
- Fixed test hang on Fedora CI: stop_git() looped forever when ps was unavailable because bash in POSIX/sh mode returns exit 0 from kill with an empty process group argument. Fixed by falling back to /proc/$pid/stat for process group ID and guarding stop_git against empty pgid.
- Redirect spawn_daemon() stdout/stderr to /dev/null and close inherited file descriptors to prevent the intermediate process from holding test pipe file descriptors
- Call setsid() on daemon detach to prevent shells with job control from waiting on the daemon process group
- Close inherited file descriptors 3-7 in the test watchdog subprocess
- Added 30-second timeout to "fsmonitor--daemon stop" to prevent indefinite blocking
- Added helpful error message when inotify watch limit (max_user_watches) is reached
- Initialize fd_inotify to -1 and use fd >= 0 check for correct fd 0 handling
- Use sysconf(_SC_OPEN_MAX) instead of hardcoded 1024 for fd close limit
- Check setsid() return value

Changes since v4:
- Added Meson build support

Changes since v3:
- Fix crash on rapid nested directory creation (EEXIST from inotify_add_watch with IN_MASK_CREATE)
- Extensive stress testing

Changes since v2:
- Fix khash memory leak in do_handle_client

Changes since v1:
- Fix hashmap memory leak in fsmonitor_run_daemon()

CC: Patrick Steinhardt <ps@pks.im>
cc: Paul Tarjan <paul@paultarjan.com>
cc: "Paul Tarjan via GitGitGadget" <gitgitgadget@gmail.com>
cc: SZEDER Gábor <szeder.dev@gmail.com>
cc: Jeff King <peff@peff.net>
cc: Ben Knoble <ben.knoble@gmail.com>

### CODE REVIEW

**Comment by ptarjan** [comment]
Sorry for the noise. Is your run with my v3 code? That one had a green CI
run. There was an existing leak that I sent as another patch as well as
bundled into this as v3 as I’m not sure of your procedure for stacked
patches.

**Comment by ptarjan** [comment]
On Tue, Feb 24, 2026 at 2:03 AM gitgitgadget-git[bot] <
***@***.***> wrote:

> *gitgitgadget-git[bot]* left a comment (git/git#2147)
> <https://github.com/git/git/pull/2147#issuecomment-3950185142>
>
> *Patrick Steinhardt* wrote on the Git mailing list
> ***@***.***> (how to reply to
> this email <https://gitgitgadget.github.io/reply-to-this>):
>
>  Thanks for the thorough review. v6 is a 10-patch series now.

> I think other than stating that this uses inotify, what this commit
> message should also explain is why it was chosen over its alternatives
> like fanotify.

Yeah, good call. fanotify needs CAP_SYS_ADMIN or CAP_PERFMON so it's a
non-starter for an unprivileged daemon. Added that.

> It would be great to avoid all these bulleted lists here and instead
> have some paragraphs that explain the design decisions.

Rewrote.

> The issues that you've addressed don't typically belong into the commit
> message. They are useful context when part of a cover letter, but are
> less useful in the commit messages themselves.

Moved to cover letter.

> In that case it might make sense to say "Based-on-patch-by:" for both
> of these authors.

Done.

> These feel like while-at-it memory leak fixes. They should probably be
> moved into separate commits with a proper explanation.

Split out as [1/10] and [2/10].

> This is (almost) the exact same implementation as we have on macOS.
> We should probably deduplicate the logic.

Yep. [5/10] merges them into fsm-ipc-unix.c. The unified version picks up
the worktree NULL BUG() guard that was only in the Linux version.

> This file is again an almost exact copy of what we have in
> "fsm-settings-darwin.c", and as far as I can see there isn't even
> anything specific to either of the systems here. So we should probably
> deduplicate the logic.

[6/10] merges them into fsm-settings-unix.c. Kept the xstrdup()+dirname()
approach from macOS and the "vfat" check.

> From these we should really only require <sys/inotify.h>, as all
> others are included via "compat/posix.h"

Removed. Also dropped the redundant <errno.h>/<stdio.h>/<string.h> from
fsm-path-utils-linux.c while I was at it.

> Error messages should start with a lower-case letter.

Fixed.

> Is it intended that the polling timeout is 1000 seconds, or ~16
> minutes? If so, it feels like something that might warrant a comment.

Nope, just a bug. Both upstream PRs had 1000 too so I inherited it without
looking closely. The stale rename check is now 1 second. I also bumped the
poll() timeout from 1ms to 50ms to reduce idle wake-ups.

> Hm. In case the watch already exists, can't it be the case that we
> have raced e.g. during a rename? For example, the old watch hasn't
> been removed yet, but we already try to create the new watch. If we
> then see EEXIST we wouldn't update the hash map and thus keep the old
> path intact.
>
> Or is the order of inotify events guaranteed?

Renames can't hit EEXIST because rename_dir() does remove_watch() before
add_watch(), so the old watch is gone by the time we add the new one.

Where EEXIST actually comes up is the recursive scan overlap. Say we get
IN_CREATE for a new directory: we add_watch() it and then
register_inotify() recursively scans its subdirectories. Meanwhile the
kernel has queued IN_CREATE events for those same subdirs. When we get back
to processing those queued events, we try add_watch() on directories that
register_inotify() already took care of. With IN_MASK_CREATE the duplicate
gets EEXIST, which is fine since the watch is already set up correctly.

> Is it correct to conflate ENOENT and ENOTDIR here? In the first case
> the directory was deleted, sure. But in the second case the directory
> might have turned into a file due to renames, so don't we have to
> treat it a bit differently?

Either way it's not a directory we can watch, and the parent's inotify
watch already delivered the relevant events (IN_MOVED_FROM/IN_MOVED_TO or
IN_DELETE/IN_CREATE). We're just doing a recursive scan here, so if the
path is gone or turned into a file, there's nothing left to scan.

> We don't use fstatat(3p) in our codebase yet, also because it's not
> easily portable to Windows. But I wonder whether we should use it here
> to be safer against races.

Yeah it'd be marginally safer, but since it's not used elsewhere and the
parent watch catches any replacement anyway, I'd rather not introduce it
here. Happy to revisit if it becomes a real problem.

> Curious. Don't we have to unregister the watcher in case a directory
> was deleted?

We do, just not in that code path. The kernel sends IN_DELETE to the parent
watch (that's what we see in process_event) and separately sends
IN_DELETE_SELF to the directory's own watch. handle_events() catches
IN_DELETE_SELF via em_remove_watch() → remove_watch(). So the break in
process_event just records the path change; the actual cleanup happens when
IN_DELETE_SELF arrives.

> Can it happen that events arrive out-of-order so that we have some
> events queued up that would touch the same path? In such a case we
> might silently ignore such queued events.

Events for a given wd come in order within a read() buffer. If we remove a
watch and then see stale events for that wd in a later read(), the hashmap
lookup returns NULL and we skip them. That's fine since the path change was
already reported when we processed IN_DELETE_SELF.

> Do we need to use a monotonic clock here to ensure that there cannot
> be any backwards jumps in time, e.g. when switching from summer to
> winter time?

time(NULL) is UTC so DST doesn't matter. An NTP step could cause a small
jump but the worst case is the stale rename check fires a bit early or
late, which is harmless. The rest of fsmonitor uses time(NULL) for batch
pinning too so this stays consistent.

> Do we have to synchronize access to `state->listen_data->shutdown`?
> As far as I can see it's being set by the fs event listener.

It's written by the listener thread and checked in that same thread's poll
loop. The only other writer is fsm_listen__stop_async() from the main
thread, and it only ever transitions from CONTINUE → terminal. Worst case
of a torn read is one extra poll iteration. The macOS and Windows backends
do the same.

> In which cases do we need to have this fallback for statfs? This
> syscall exists in Linux since the 90s, so shouldn't we be able to
> assume that we can use it? Or are there specific error cases that we
> need to worry about here?

It's not a fallback for statfs. statfs() always works and gives us the fs
magic number for the is_remote check. find_mount() reads /proc/mounts to
get a human-readable fstype string ("nfs4" instead of 0x6969) for trace
logging. If /proc/mounts isn't available we just fall back to a
magic-number-to-name lookup table.

> It's a bit unclear why specifically this test was added. Does it catch
> an edge case that you have discovered? Might make sense to also add it
> in a preparatory commit so that we can get a bit of context.

Yeah, this caught an actual crash in v3 (the EEXIST race I described
above). Moved to its own commit [8/10] with the context in the message.

Four new patches in this version:

[3/10] and [4/10]: pthread_cond_timedwait. [3/10] adds
pthread_cond_timedwait to the Windows pthread compat layer, [4/10] uses it
to replace the infinite pthread_cond_wait() in the cookie-wait path with a
1-second timeout. On overlay filesystems (common in containers) inotify
watches succeed but events never arrive, so without this the daemon blocks
forever waiting for the cookie event.

[9/10] and [10/10]: fd cleanup and daemon detach. [9/10] adds a
close_fd_above_stderr option to child_process that closes fds 3+ after fork
(uses sysconf(_SC_OPEN_MAX) capped at 4096). [10/10] uses it in
spawn_daemon() and also calls setsid() on --detach so shells with job
control don't wait on the daemon's process group. This is what was causing
the CI hang on Fedora. stop_git() also had a bug where `ps` wasn't
available (no procps-ng), leaving git_pgid empty, and bash in sh mode
silently accepts `kill -0 -- -` so the loop ran forever.

File: builtin/fsmonitor--daemon.c [hunk 1]
```diff
@@ -197,20 +197,31 @@ static enum fsmonitor_cookie_item_result with_lock__wait_for_cookie(
 	unlink(cookie_pathname.buf);
```
[the remark this answers is not in the document]
Comment (ptarjan) [reply builtin/fsmonitor--daemon.c:207]: The 1-second timeout only fires when the filesystem fails to deliver the cookie event at all (e.g. overlayfs in containers where inotify watches succeed but events never arrive). On a working filesystem the cookie event comes back in well under a millisecond, so the timeout never triggers. When it does fire, the client falls back to a full scan, which is the safe default. Happy to make it configurable if you think that's worth it, but the current behavior seemed reasonable as a starting point.

[the remark this answers is not in the document]
Comment (ptarjan) [reply builtin/fsmonitor--daemon.c:207]: Fixed, changed to "observe".

File: Makefile [hunk 2]
```diff
@@ -2365,7 +2365,11 @@ ifdef FSMONITOR_DAEMON_BACKEND
 	COMPAT_CFLAGS += -DHAVE_FSMONITOR_DAEMON_BACKEND
```
[the remark this answers is not in the document]
Comment (ptarjan) [reply Makefile:2365]: Good point. Restructured: FSMONITOR_OS_SETTINGS is now set to "unix" for both macOS and Linux (and "win32" for Windows). The build files use $(FSMONITOR_OS_SETTINGS) for fsm-ipc and fsm-settings, and $(FSMONITOR_DAEMON_BACKEND) for the platform-specific files (listen, health, path-utils). No more if/else blocks.

File: Makefile [hunk 3]
```diff
@@ -2365,12 +2365,20 @@ ifdef FSMONITOR_DAEMON_BACKEND
...
 	COMPAT_OBJS += compat/fsmonitor/fsm-listen-$(FSMONITOR_DAEMON_BACKEND).o
 	COMPAT_OBJS += compat/fsmonitor/fsm-health-$(FSMONITOR_DAEMON_BACKEND).o
-	COMPAT_OBJS += compat/fsmonitor/fsm-ipc-$(FSMONITOR_DAEMON_BACKEND).o
+ifeq ($(FSMONITOR_DAEMON_BACKEND),win32)
+	COMPAT_OBJS += compat/fsmonitor/fsm-ipc-win32.o
+else
+	COMPAT_OBJS += compat/fsmonitor/fsm-ipc-unix.o
+endif
 endif
```
[the remark this answers is not in the document]
Comment (ptarjan) [reply Makefile:2369]: Yep, same fix applied here. Also rewrote both commit messages to describe what's actually happening (renaming darwin to unix) rather than "merging two implementations".

File: builtin/fsmonitor--daemon.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (dscho) [reply builtin/fsmonitor--daemon.c:207]: @ptarjan please note that https://github.com/gitgitgadget/gitgitgadget/issues/154 is not implemented, and therefore all of your replies here in the PR are not reaching their intended audience. You will need to follow the guidance in the "reply to this" links.

[the remark this answers is not in the document]
Comment (ptarjan) [reply builtin/fsmonitor--daemon.c:207]: Thanks for the heads up. I was mostly doing this for self tracking to make sure I addressed them all. Is the changelog in the PR body enough of a reply or should I forward all these notes to the list by hand?

[the remark this answers is not in the document]
Comment (dscho) [reply builtin/fsmonitor--daemon.c:207]: > Is the changelog in the PR body enough of a reply or should I forward all these notes to the list by hand?

It's up to you. In the past I have skipped the personal replies when under enormous time pressure. I feel that this did not go over well, even if I did not have much choice on the matter.

**Comment by ptarjan** [comment]
On Wed, Mar 4, 2026, Junio C Hamano wrote:
> I wonder if a generic callback function
> to call here in the child between fork and exec that the caller can
> supply would be a good thing to have.

Done in v9. Replaced the close_fd_above_stderr flag with a pre_exec_cb
function pointer on struct child_process. The fd-closing logic is now
a standalone close_fd_above_stderr() function that the two fsmonitor
callers pass as the callback.
